src/hotspot/share/opto/mulnode.cpp

```diff
@@ -201,6 +201,18 @@ const Type* MulNode::Value(PhaseGVN* phase) const {
 //------------------------------Ideal------------------------------------------
 // Check for power-of-2 multiply, then try the regular MulNode::Ideal
 Node *MulINode::Ideal(PhaseGVN *phase, bool can_reshape) {
+  // convert "(-a)*(-b)" into "a*b"
+  Node *in1 = in(1);
+  Node *in2 = in(2);
+  if (in1->Opcode() == Op_SubI && in2->Opcode() == Op_SubI) {
+    Node* n11 = in1->in(1);
+    Node* n21 = in2->in(1);
+    if (phase->type(n11)->higher_equal(TypeInt::ZERO) &&
+        phase->type(n21)->higher_equal(TypeInt::ZERO)) {
```

Comment:
I was thinking if it's a good idea to move these code into MulNode, as they were actually much the same with MulLNode.

Reply:
I wonder that too, so is the rest of MulINode/MulLNode::Ideal() code (and many other places). I am not sure how to workaround the different types, any suggestions?

Reply:
Just a dogfood, but it works. https://gist.github.com/theRealELiu/328d62157975b1f20e3626b3ef747eb4

Too much abstraction makes the code hard to read. One needs to check the concrete class to identify what the code exactly is, E.g.  In my patch, add_id() may be TypeInt::ZERO or TypeLong::Zero, even TypeD::ZERO.  So I'm not sure if it's a good idea. Is there any guidelines to this issue, try to abstract them or make the readability in the first place? @TobiHartmann

Reply:
Yes, I would also prefer to move the optimization into `MulNode::Ideal`. @theRealELiu's patch is good but can be further improved by modifying the node inputs instead of returning a new node (similar to the other optimizations in `MulNode::Ideal`).

Reply:
Also, `Type::is_zero_type` can be used to detect 0 and instead of checking the opcodes, `Node::is_Sub` should be used.

Reply:
Nice! Thanks, I will make changes accordingly.